### DISCUSSION TITLE
Deploy Pimcore class definitions

### DISCUSSION THREAD

**Discussion by head1328** [pimcore/pimcore#10576] [Q&A / Support]
I have some questions about this: https://pimcore.com/docs/pimcore/current/Development_Documentation/Deployment/Deployment_Tools.html#page_Pimcore-Class-Definitions

* after creating some classes on my local machine with dev-mode on and `PIMCORE_CLASS_DEFINITION_WRITABLE=true`, I copied the definition files from `/var/classes/definitions-*.php` to `/config/pimcore/classes/`
* because of the .gitignore file I committed both folders with the same definition classes, but without DataObjects
* on my remote machine without `PIMCORE_CLASS_DEFINITION_WRITABLE` I executed `pimcore:deployment:classes-rebuild -c` and this results in: `In ClassDefinition.php line 442: definitions in config/pimcore folder cannot be overwritten`

How do you understand that in the documentation and how exactly does that work?

**Comment by dvesh3** [comment]
@head1328 you need to set `PIMCORE_CLASS_DEFINITION_WRITABLE=true` env variable on remote machine to rebuild classes. I know the language in the documentation is misleading as it starts with "For your development environment .." so i have kicked it https://github.com/pimcore/pimcore/commit/ddc091dd0449505fcbbfd8c59b2801cb675b6c38

**Reply by head1328** [reply]
OK, I need this set to true for rebuilding but why? When it comes to rebuilding, we don't speak about updating any class definitions, it's just about creating DataObjects from class definitions. Sorry but I don't understand this "feature".

When I copy my class definitions to `/config/pimcore/classes`, do I still have to commit & push the class definitions under the `var/classes` folder?

**Reply by dvesh3** [reply]
The feature is about changing the path for class definitions to `/config/pimcore/classes` and it works the same way as for `var/classes`. Additionally, it needs writeable permissions to create dataobject classes under the same path. please see https://github.com/pimcore/pimcore/issues/7980

**Reply by head1328** [reply]
Sorry but, again, this is not working as expected:

At first I created 3 classes and the `defintion_*.php` files + `DataObjects`  are located under `/var/classes` as expected (default behavior). Then I moved the definition files to `config/pimcore/classes` (without DataObjects) because I want to ignore the entire var folder when it comes to deploying the application later. With `PIMCORE_CLASS_DEFINITION_WRITABLE=true` (I tried with export inside my docker container, .env file and config/pimcore/constants.php) is not working. The command `pimcore:deployment:classes-rebuild` complains with `Warning: file_put_contents(/config/pimcore/classes/DataObject/Category.php): Failed to open stream: No such file or directory` and with the `-c` option nothing happend as well.
